cpp/src/arrow/type_traits.h

```diff
@@ -244,6 +260,8 @@ struct TypeTraits<UnionType> {
 template <>
 struct TypeTraits<DictionaryType> {
   using ArrayType = DictionaryArray;
+  // TODO(wesm): Not sure what to do about this
```

Comment:
It seems like maybe one shouldn't exist?  Otherwise just an int and a dictionary pointer?

Reply:
We're also lacking BuilderType here for fundamentally the same reason. `pair<int64_t, std::shared_ptr<Array>>` is attractive but is it acceptable for when the index type is `uint64_t`?

An option is to use/wrap `Scalar` or `std::shared_ptr<Scalar>` here. Then BuilderType could be TypeErasedDictionaryBuilder or so:

```
std:: unique_ptr<ArrayBuilder> builder;
RETURN_NOT_OK(MakeBuilder(pool, dictionary(int32(), string_array), &builder));
for (std::shared_ptr<Scalar> s : user_inputs) {
  RETURN_NOT_OK(builder->Append(s));
  // Append null for strings not in string_array
  // and non-StringScalars
}
```

Reply:
I don't think a decision needs to be made right now. We need to write some kernels that involve dictionaries first